docs/settings.md
# Settings

All settings are defined through environment variable.

## Global configuration

| Name              | Type    | Default           | Comment                      |
| ----------------- | ------- | ----------------- | ---------------------------- |
| VERSION           | String  | 0.0.0-development | Application version          |
| APP_PUBLIC_PATH   | String  | /public/          | Public path for assets       |
| APP_PORT          | Number  | 3000              | HTTP server listening port   |
| APP_GZIP          | Boolean | true              | Enable GZip compression      |
| APP_SECURE_COOKIE | Boolean | true              | Enable secure cookies        |
| APP_COOKIE_POLICY | String  | strict            | Cookie sameSite policy       |
| APP_INTROSPECTION | Boolean |                   | Enable GraphQL introspection |
| APP_VERBOSE       | Boolean | true              | Enable console logs          |

It's recommended to disable compression if you may delegate it to a reverse proxy.

## Database configuration

| Name        | Type   | Default                       | Comment                    |
| ----------- | ------ | ----------------------------- | -------------------------- |
| APP_DB_URI  | String | mongodb+srv://localhost:27017 | Database connection string |
| APP_DB_NAME | String | app                           | Database Name              |
| APP_DB_POOL | Number | 10                            | Database pool size         |

## Database encryption configuration

| Name                       | Type    | Default                 | Comment                                       |
| -------------------------- | ------- | ----------------------- | --------------------------------------------- |
| APP_CSFLE_MODE             | String  |                         | CSFLE provider (supported: aws, local)        |
| APP_CSFLE_KEY_VAULT        | String  | encryption.\_\_keyVault | Key Vault                                     |
| APP_CSFLE_MASTER_KEY       | String  |                         | Local path to the master key (local provider) |
| APP_CSFLE_ACCESS_KEY       | String  |                         | AWS Access key ID                             |
| APP_CSFLE_SECRET_KEY       | String  |                         | AWS Access secret key                         |
| APP_DB_CRYPTD_URI          | String  |                         | MongoCryptd URI                               |
| APP_DB_CRYPTD_BYPASS_SPAWN | Boolean | false                   | Bypass MongoCryptd Spwan                      |
| APP_DB_CRYPTD_SPWAN_ARGS   | String  |                         | Extra arguments for MongoCryptd Spwan         |

## Redis configuration

| Name          | Type   | Default                | Comment                 |
| ------------- | ------ | ---------------------- | ----------------------- |
| APP_REDIS_URI | String | redis://127.0.0.1:6379 | Redis connection string |

## Health checks configuration

| Name                   | Type    | Default             | Comment                               |
| ---------------------- | ------- | ------------------- | ------------------------------------- |
| APP_HEALTH_ENABLED     | Boolean | false               | Serve health checks                   |
| APP_HEALTH_PORT        | Number  | 4000                | Port on which serve the health checks |
| APP_HEALTH_ALLOWED     | String  | ::1/128,127.0.0.0/8 | Allowed CIDR for the health checks    |
| APP_HEALTH_WORKER_BEAT | String  |                     | URL for worker beat pushes            |

## Session configuration

| Name                 | Type   | Default | Comment                                        |
| -------------------- | ------ | ------- | ---------------------------------------------- |
| APP_SESSION_SECRET   | String |         | Secret use for JWT signature on session tokens |
| APP_SESSION_LIFETIME | String | 1h      | Session token lifetime                         |

## SMTP configuration

| Name              | Type    | Default               | Comment                           |
| ----------------- | ------- | --------------------- | --------------------------------- |
| APP_SMTP_HOST     | String  | localhost             | SMTP host                         |
| APP_SMTP_PORT     | Number  | 465                   | SMTP Port                         |
| APP_SMTP_SECURE   | Boolean | false                 | SMTP transport on secure mode     |
| APP_SMTP_USER     | String  |                       | SMTP user                         |
| APP_SMTP_PASSWORD | String  |                       | SMTP password                     |
| APP_SMTP_FROM     | String  | noreply@appvantage.co | Default sender on outgoing emails |

## Storage configuration

| Name                   | Type    | Default        | Comment                                  |
| ---------------------- | ------- | -------------- | ---------------------------------------- |
| APP_STORAGE_ENDPOINT   | String  |                | API endpoint for the Object storage      |
| APP_STORAGE_ACCESS_KEY | String  |                | Access Key for object storage            |
| APP_STORAGE_SECRET_KEY | String  |                | Secret Key for object storage            |
| APP_STORAGE_SSL        | Boolean | true           | Use TLS with the object storage          |
| APP_STORAGE_PORT       | Number  |                | Port on which the API endpoint is served |
| APP_STORAGE_REGION     | String  | ap-southeast-1 | Object storage region                    |
| APP_STORAGE_BUCKET     | String  | app            | Bucket name for the object storage       |

## Sentry configuration

| Name                          | Type    | Default | Comment                               |
| ----------------------------- | ------- | ------- | ------------------------------------- |
| APP_SENTRY_DSN                | String  |         | Sentry DSN                            |
| APP_SENTRY_RELEASE            | String  |         | Sentry version                        |
| APP_SENTRY_ENVIRONMENT        | String  |         | Sentry environment                    |
| APP_SENTRY_TRACING            | Boolean | true    | Enable Sentry tracing                 |
| APP_SENTRY_TRACES_SAMPLE_RATE | Number  | 1.0     | Sentry traces sample rate for tracing |

## HTML2PDF configuration

| Name                  | Type   | Default | Comment                       |
| --------------------- | ------ | ------- | ----------------------------- |
| APP_HTML2PDF_ENDPOINT | String |         | Endpoint for the html2pdf API |

## Limiter configuration

| Name            | Type   | Default | Comment                                   |
| --------------- | ------ | ------- | ----------------------------------------- |
| APP_LIMITER_API | Number | 1000    | Rate limiter per second per IP on the API |


src/server/core/config.ts
import chalk from 'chalk';
import type { SMTPSettings } from '../emails';
import { getClientSideFieldLevelEncryptionSettings } from './encryption';
import { getString, getBoolean, getInteger, getNumber, getPrefix, getStringList } from './env';

const getSmtpSettings = (): SMTPSettings => {
    const base = {
        host: getString(getPrefix('SMTP_HOST'), 'localhost'),
        port: getInteger(getPrefix('SMTP_PORT'), 465),
        secure: getBoolean(getPrefix('SMTP_SECURE'), false),
    };

    const user = getString(getPrefix('SMTP_USER'));

    if (!user) {
        return base;
    }

    return {
        ...base,
        auth: {
            user,
            pass: getString(getPrefix('SMTP_PASSWORD')),
        },
    };
};

const version = getString('VERSION', '0.0.0-development');

const config = {
    version,

    // miscellaneous
    verbose: getBoolean(getPrefix('VERBOSE'), true),

    // introspection state on the GraphQL API
    introspection: getBoolean(getPrefix('INTROSPECTION'), process.env.NODE_ENV !== 'production'),

    port: getNumber(getPrefix('PORT'), 3000),
    publicPath: getString(getPrefix('PUBLIC_PATH'), '/public/'),

    // secure cookies
    cookiePolicy: getString(getPrefix('COOKIE_POLICY'), 'strict'),
    secureCookie: getBoolean(getPrefix('SECURE_COOKIE'), true),

    // health checks
    healthChecks: {
        enabled: getBoolean(getPrefix('HEALTH_ENABLED'), false),
        port: getNumber(getPrefix('HEALTH_PORT'), 4000),
        allowed: getStringList(getPrefix('HEALTH_ALLOWED'), ['::1/128', '127.0.0.0/8']),
        workerBeat: getString(getPrefix('HEALTH_WORKER_BEAT')),
    },

    // gzip module
    gzip: getBoolean(getPrefix('GZIP'), true),

    // internationalization
    i18n: {
        locales: ['en'],
    },

    // server runtime
    db: {
        uri: getString(getPrefix('DB_URI'), 'mongodb+srv://localhost:27017'),
        name: getString(getPrefix('DB_NAME'), 'app'),
        pool: getInteger(getPrefix('DB_POOL'), 10),

        // Client Side Field Level Encryption (CSFLE)
        // https://docs.mongodb.com/drivers/security/client-side-field-level-encryption-guide/
        encryption: getClientSideFieldLevelEncryptionSettings(),

        // mongocryptd settings
        // only used if encryption is enabled
        cryptd: {
            uri: getString(getPrefix('DB_CRYPTD_URI')),
            mongocryptdBypassSpawn: getBoolean(getPrefix('DB_CRYPTD_BYPASS_SPAWN'), false),
            mongocryptdSpawnArgs: getStringList(getPrefix('DB_CRYPTD_SPWAN_ARGS'), [], ' '),
        },
    },

    redis: {
        uri: getString(getPrefix('REDIS_URI'), 'redis://127.0.0.1:6379'),
    },

    session: {
        secret: getString(getPrefix('SESSION_SECRET')),
        lifetime: getString(getPrefix('SESSION_LIFETIME'), '1h'),
    },

    // default/system SMTP settings
    smtp: {
        transporter: getSmtpSettings(),
        sender: getString(getPrefix('SMTP_FROM'), 'noreply@appvantage.co'),
    },

    storage: {
        provider: {
            endPoint: getString(getPrefix('STORAGE_ENDPOINT')),
            accessKey: getString(getPrefix('STORAGE_ACCESS_KEY')),
            secretKey: getString(getPrefix('STORAGE_SECRET_KEY')),
            useSSL: getBoolean(getPrefix('STORAGE_SSL'), true),
            port: getInteger(getPrefix('STORAGE_PORT')),
            region: getString(getPrefix('STORAGE_REGION'), 'ap-southeast-1'),
        },
        bucket: getString(getPrefix('STORAGE_BUCKET'), 'app'),
    },

    sentry: {
        dsn: getString(getPrefix('SENTRY_DSN')),
        release: getString(getPrefix('SENTRY_RELEASE')),
        environment: getString(getPrefix('SENTRY_ENVIRONMENT')),
        tracing: getBoolean(getPrefix('SENTRY_TRACING'), true),
        tracesSampleRate: getNumber(getPrefix('SENTRY_TRACES_SAMPLE_RATE'), 1.0),
    },

    html2pdf: {
        endpoint: getString(getPrefix('HTML2PDF_ENDPOINT')),
    },

    limiter: {
        api: getNumber(getPrefix('LIMITER_API'), 1000),
    },

    prometheus: {
        enabled: getBoolean(getPrefix('PROMETHEUS_ENABLED'), false),
        internal: getBoolean(getPrefix('PROMETHEUS_INTERNAL'), false),
        internalPort: getInteger(getPrefix('PROMETHEUS_INTERNAL_PORT'), 7788),
        external: getBoolean(getPrefix('PROMETHEUS_EXTERNAL'), false),
        externalPath: getString(getPrefix('PROMETHEUS_EXTERNAL_PATH'), '/metrics'),
        prefix: getString(getPrefix('PROMETHEUS_PREFIX'), 'app_'),
    },
};

export type RuntimeConfig = typeof config;

const exitOnError = (error: string): void => {
    console.error(error);
    process.exit(1);
};

export const runValidityChecks = () => {
    if (!config.session.secret) {
        exitOnError(chalk.red('APP_SESSION_SECRET is missing in environment variables'));
    }
};

export default config;
